vm/src/object/core.rs

```diff
@@ -121,6 +204,49 @@ struct PyInner<T> {
     payload: T,
 }
 
+#[cfg(feature = "gc")]
+unsafe impl Trace for PyInner<Erased> {
+    fn trace(&self, tracer_fn: &mut TracerFn) {
+        // trace PyInner's other field(that is except payload)
+        // self.typ.trace(tracer_fn);
+        self.dict.trace(tracer_fn);
+        // weak_list keeps a *pointer* to a struct for maintaince weak ref, so no ownership, no trace
+        self.slots.trace(tracer_fn);
+
+        /// FIXME(discord9): Optional trait bound(Like a ?GcTrace) require specialization
+        ///
+        /// https://stackoverflow.com/questions/68701910/function-optional-trait-bound-in-rust
+        ///
+        /// fall back to use TypeId for now
+        macro_rules! optional_trace {
+            ($($TY: ty),*$(,)?) => {
+                $(
+                    if TypeId::of::<$TY>() == self.typeid{
+                        // Safety: because typeid said so!
+                        let inner: &PyInner<$TY> = unsafe { &*(self as *const PyInner<Erased> as *const PyInner<$TY>) };
+                        inner.payload.trace(tracer_fn);
+                    }
+                )else*
+            };
+        }
```

Comment:
> Hmm... what if instead of this, `PyObjectPayload` required Trace (and stuck it in the PyObjVtable), but there's a DummyTrace trait that's safe and marks a object that isn't actually Trace. like:
> 
> ```rust
> unsafe trait Trace {
>     const IS_TRACE: bool = true;
>     fn trace(&self, ...);
> }
> trait DummyTrace {} // or NoTrace or something
> unsafe impl<T: DummyTrace> Trace for T {
>     const IS_TRACE: bool = false;
>     fn trace(&self, ...) {}
> }
> ```
> 
> And `#[pyclass]` could by default add a DummyTrace impl, unless specifically asked to pytrace it.
> 
> Oh and instead of `TraceHelper::is_traceable` the vtable has a `trace: Option<...>` that's None if IS_TRACE is false. Might be faster too, rather than checking 70 different TypeIds it's just one virtual function call.

Interesting idea, will try refactor into that

Reply:
> 

Might rename those to `MaybeTrace` `try_trace` and make `is_trace` a const function